espei/paramselect.py

```diff
@@ -1503,107 +645,91 @@ def fit(input_fname, datasets, resume=None, scheduler=None, recfile=None,
     phase_models = dask.delayed(phase_models, pure=True)
-    dbf, phase_models = \
-        scheduler.persist([dbf, phase_models], broadcast=True)
-
-    error_args = ",".join(['{}=model_dof[{}]'.format(x, idx) for idx, x in
-                           enumerate(symbols_to_fit)])
-    error_code = """
-    def error({0}):
-        parameters = OrderedDict(sorted(locals().items(), key=str))
-        import time
-        enter_time = time.time()
-        try:
-            iter_error = multi_phase_fit(dbf, comps, phases, datasets, phase_models,
-                                         parameters=parameters, scheduler=scheduler)
-        except ValueError as e:
-            #print(e)
-            iter_error = [np.inf]
-        iter_error = [np.inf if np.isnan(x) else x**2 for x in iter_error]
-        iter_error = -np.sum(iter_error)
-        if recfile:
-            print(time.time()-enter_time, 'exit', iter_error, flush=True)
-            recfile.write(','.join([str(-iter_error), str(time.time()-enter_time)] + [str(x) for x in parameters.values()]) + '\\n')
-        return iter_error
-    """
-    import textwrap
-    error_code = textwrap.dedent(error_code).format(error_args)
-    if recfile:
-        recfile.write(','.join(
-            ['error', 'time'] + [str(x) for x in symbols_to_fit]) + '\n')
 
-    result_obj = {'model_dof': model_dof}
-    error_context = {'data': data, 'comps': comps, 'dbf': dbf,
-                     'phases': sorted(data['phases'].keys()),
+    # contect for the log probability function
+    error_context = {'data': data, 'comps': comps, 'dbf': dbf, 'phases': sorted(data['phases'].keys()),
                      'datasets': datasets, 'symbols_to_fit': symbols_to_fit,
-                     'phase_models': phase_models, 'scheduler': scheduler,
-                     'recfile': recfile}
-    error_context.update(globals())
-    exec(error_code, error_context, result_obj)
-    error = result_obj['error']
-    error = pymc.potential(error)
-    model_dof.append(error)
-    pymod = pymc.Model(model_dof)
-
-    def close_mdl():
-        mdl.db.close()
-        if recfile:
-            recfile.close()
-
-    if tracefile is not None:
-        mdl = pymc.MCMC(pymod, db='txt', dbname=tracefile)
-    else:
-        mdl = pymc.MCMC(pymod)
+                     'phase_models': phase_models, 'recfile': recfile}
+
+
+    def save_tracefile(sampler):
+        if tracefile:
+            logging.debug('Writing chain to {}'.format(tracefile))
+            np.savetxt(tracefile, sampler.flatchain)
+
+    # set up the MCMC run
+    # set up the initial parameters
+    # apply a Gaussian random to each parameter with std dev of 0.10*parameter
+    initial_parameters = np.array(initial_parameters)
+    ndim = len(initial_parameters)
+    nwalkers = 2*ndim # walkers must be of size (2n*ndim)
+    initial_walkers = np.tile(initial_parameters, (nwalkers, 1))
+    walkers = np.random.normal(initial_walkers, np.abs(initial_walkers*0.10))
```

Comment:
To try to limit stochasticity for debugging purposes I would use
```python
rng = np.random.RandomState(1769)
# ...
walkers = rng.normal(...)
```

I know emcee has its own source of randomness, but that can be addressed later.